### PR TITLE
Feature raster tiles

### PR DESCRIPTION
**Pull request by Quillasp**
This is the first step of the raster tiles rendering feature #71. There's:

- Raster style layer from the [raster](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#raster) layer mapbox specification,
- Raster layer that will hold the raster data
- Two differents sources for vector tiles (Tessellate source) and raster source that will load different data following the layer type.

## ✔️ PR Todo

- [x] Discussing how the different layer types should be fetched (one source client for the vector tiles and another one for the raster tiles? or fetching from the same source client with different address based on the layer type)
- [x] Raster data loading and storing
- [ ] Raster rendering

### CODE REVIEW

File: maplibre/src/io/tile_repository.rs [hunk 1]
```diff
@@ -116,22 +124,45 @@ impl TileRepository {
...
+    //             if !tessellated_set.contains(layer.as_str()) {
+    //                 return true;
+    //             }
+    //         }
+
+    //         return false;
+    //     }
+    //     true
+    // }
+    pub fn is_layers_missing(
```
Comment (ybiletskyi) [inline maplibre/src/io/tile_repository.rs:145]: Probably rename to `missed_layers`. Clean up the commented code.

File: maplibre/src/stages/request_stage.rs [hunk 2]
```diff
@@ -135,26 +213,33 @@ where
...
-                                    .unwrap(),
-                                Err(e) => {
-                                    log::error!("{:?}", &e);
-                                    state.tile_unavailable(&coords, request_id).unwrap()
-                                }
-                            }
-                        })
-                    }))
-                    .unwrap();
+                // let client = SourceClient::Http(self.http_source_client.clone());
```
Comment (ybiletskyi) [inline maplibre/src/stages/request_stage.rs:216]: Clean up commented code.

File: maplibre/src/style/style.rs [hunk 3]
```diff
@@ -129,6 +130,17 @@ impl Default for Style {
...
                     source_layer: Some("boundary".to_string()),
                 },
+                StyleLayer {
+                    index: 0,
+                    id: "raster".to_string(),
+                    typ: "raster".to_string(),
+                    maxzoom: None,
+                    minzoom: None,
+                    metadata: None,
+                    paint: Some(LayerPaint::Raster(RasterLayer::default())),
```
Comment (ybiletskyi) [inline maplibre/src/style/style.rs:140]: `paint` should be `None`.

File: maplibre/src/style/layer.rs [hunk 4]
```diff
@@ -39,17 +66,20 @@ pub enum LayerPaint {
     Line(LinePaint),
     #[serde(rename = "fill")]
     Fill(FillPaint),
+    #[serde(rename = "raster")]
```
Comment (ybiletskyi) [inline maplibre/src/style/layer.rs:69]: Clean up this changes. We will not use `LayerPaint` for raster tiles.

**Review by ybiletskyi — CHANGES_REQUESTED** [review]

File: maplibre/src/style/layer.rs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 //! Vector tile layer drawing utilities.
```
Comment (ybiletskyi) [inline maplibre/src/style/layer.rs:2]: Looks like changes in this file is not used. Please clean up.

File: maplibre-demo/src/main.rs [hunk 6]
```diff
@@ -67,8 +68,16 @@ fn run_headless() {
         let coords = WorldTileCoords::from((0, 0, ZoomLevel::default()));
         let request_id = 0;
 
+        let tile_coords = coords.into_tile(TileAddressingScheme::TMS).unwrap();
+        let url = format!(
+            "https://maps.tuerantuer.org/europe_germany/{z}/{x}/{y}.pbf",
+            x = tile_coords.x,
+            y = tile_coords.y,
+            z = tile_coords.z
+        );
```
Comment (maxammann) [inline maplibre-demo/src/main.rs:77]: This transformation should happen within the source client.

File: maplibre/src/io/source_client.rs [hunk 7]
```diff
@@ -47,9 +47,9 @@ impl<HC> SourceClient<HC>
 where
     HC: HttpClient,
 {
-    pub async fn fetch(&self, coords: &WorldTileCoords) -> Result<Vec<u8>, Error> {
+    pub async fn fetch(&self, url: &str) -> Result<Vec<u8>, Error> {
```
Comment (maxammann) [inline maplibre/src/io/source_client.rs:50]: The previous interface should be kept. What was the reason for the change?

File: maplibre/src/stages/request_stage.rs [hunk 8]
```diff
@@ -77,6 +79,85 @@ where
     }
 }
 
+pub trait Source<SM, HC>
```
Comment (maxammann) [inline maplibre/src/stages/request_stage.rs:82]: Source should not be part of the stages module

File: maplibre/src/stages/request_stage.rs [hunk 9]
```diff
@@ -77,6 +79,85 @@ where
...
+where
+    SM: ScheduleMethod,
+    HC: HttpClient,
+{
+    fn load(
+        request_stage: &RequestStage<SM, HC>,
```
Comment (maxammann) [inline maplibre/src/stages/request_stage.rs:88]: A source should not depend on the RequestStage and vice-versa

File: maplibre/src/style/layer.rs [hunk 10]
```diff
@@ -29,6 +30,32 @@ pub struct LinePaint {
     // TODO a lot
 }
 
+pub enum PaintType {
+    Argb(Alpha<EncodedSrgb<f32>>),
+    Raster(RasterLayer),
```
Comment (maxammann) [inline maplibre/src/style/layer.rs:35]: A raster layer is not related to the color. The a raster layer should be another case for the `LayerPaint` enum.

**Review by maxammann — CHANGES_REQUESTED** [review]

**Comment by maxammann** [comment]
@Quillasp I merged in main and fixed all conflicts

**Comment by maxammann** [comment]
@Quillasp I would propose not to try to merge main into this PR, but manually check the relevant changes and then copy them to a new branch.
A lot has changed in the past months, so there will definitely be a lot of changes.

**Comment by maxammann** [comment]
Closing in favor of #205

**Pull request closed without merging**